01-messaging.md

```diff
@@ -82,6 +83,104 @@ however, adding a 6-byte padding after the type field was considered
+the backwards-compatible addition of new fields to existing message types. This
+format describes the TLV encoding specific to wire messages.
+
+A `tlv_record` represents a single field, encoded in the form:
+
+* [`1`: `type`]
+* [`2`: `length`]
+* [`length`: `value`]
+
+A `tlv_stream` is a series of (possibly zero) `tlv_record`s, represented as the
+concatenation of the encoded `tlv_record`s. When used to extend existing
+messages, a `tlv_stream` is typically placed after all currently defined fields.
+
+The `type` is a single-byte identifier for the `tlv_record` and determines how
+the contents of `value` should be decoded. The `type` 255 is reserved for use as
+a sentinel value indicating the end of the `tlv_stream`; the length of `type`
+255 is always zero.
+
+The `length` is a big-endian 16-bit integer to signal the size in bytes of
+`value`.
+
+The `value` depends entirely on the `type`. Each `tlv_stream` has its own 8-bit
+namespace for defining the meaning of its `type` identifiers.
+
+### Requirements
+
+The sending node:
+ - MUST order `tlv_record`s in a `tlv_stream` by monotonically-increasing `type`.
+ - SHOULD NOT use the varbytes encoding for `value`s containing byte arrays.
+ - MUST NOT use `type` 255 to encode a field.
+ - MAY use `type` 255 to signal the end of a `tlv_stream`.
+   - if used, MUST set the `length` of `type` 255 to 0.
+
+The receiving node:
+ - if encounters EOF while parsing a `type`:
+   - MUST stop parsing the `tlv_stream`.
+ - if encounters EOF while parsing a `length` or `value`:
+   - MUST fail to parse to `tlv_stream`.
+ - if `type` and `length` parse:
+   - if `type` is known:
+     - MUST decode the next `length` bytes using the known encoding for `type`.
+   - if `type` is unknown:
+     - if the `tlv_stream` is not covered by a signature:
+       - MUST discard the next `length` bytes.
+     - otherwise:
+       - MUST retain a `tlv_record` for `type` using the next `length` bytes as
+         its `value`.
+ - if the `type`s in a `tlv_stream` are not monotonically-increasing:
+   - MUST fail to parse the `tlv_stream`.
+ - if the `length` of `type` 255 is not 0:
+   - MUST fail to parse the `tlv_stream`.
+
+### Rationale
+
+The primary advantage in using TLV is that a reader is able to ignore new fields
+that it does not understand, since each field carries the exact size of the
+encoded element. Without TLV, even if a node does not wish to use a particular
+field, the node is forced to add parsing logic for that field in order to
+determine the offset of any fields that follow.
+
+The sentinel value is chosen as 255 since it maintains that all records are
+sorted in ascending order, and implicitly skips over all possible `type`
+identifiers. Including a sentinel value allows, for example, concatenating
+`tlv_stream`s with different namespaces.
+
+The monotonicity constraint ensures that all `type`s are unique and can appear
+at most once.
```

Comment:
Is there a reason why you want to restrict types to appear at most once in a tlv stream? I think it could be useful and more flexible to let them appear multiple times.

Reply:
replied in main body where there was ample space 😬

Reply:
I think it could be useful to add a sentence here to explain that if multiple occurrences of the same type is needed, it should rather be done at the type definition level (so we won't need to re-discuss this when newcomers read the spec and wonder why this limitation is here).

Reply:
added in 2583ae291eaa672f12da4a35f057ee1f48f71871